File: Algorithms/DynamicProgramming/CpSolns/tribonacci-1137.cpp
// Leetcode #1137 - N-th Tribonacci Number

#include<iostream>
#include<vector>

using std::vector;
using std::cout;
using std::endl;

class Tribonacci{
    public:
        int tribonacci(int n){
            vector<int> triNums = {0, 1, 1};
            for(int i = 3; i < n+1; i++)
                triNums.push_back(triNums[i-1] + triNums[i-2] + triNums[i-3]);
            return triNums[n];
        }
};

int main(){
    Tribonacci soln;
    cout << soln.tribonacci(4) << endl;
    cout << soln.tribonacci(25) << endl;
}

File: Algorithms/DynamicProgramming/JavaSolns/Tribonacci-1137.java
// Leetcode #1137 - N-th Tribonacci number

class Tribonacci{
    static int nthTrib(int n){
        int dp[] = new int[n+1];
        if(n <= 1) return n;
        // initialize first three tribonacci numbers
        dp[0] = 0; dp[1] = 1; dp[2] = 1;
        for(int i = 3; i <= n; i++)
            // tribonacci n is sum of previous three value
            dp[i] = dp[i-1] + dp[i-2] + dp[i-3];
        // return nth trobinacci
        return dp[n];
    }
    public static void main(String[] args){
        for(int i = 20; i < 30; i++)
            System.out.println(nthTrib(i));
    }
}
